### PR TITLE
Use correct error codes and texts in CIM bundles for charges

### PR DESCRIPTION
**Pull request by BjarkeMeier**
## Description

With this pull request, we create descriptive error messages containing relevant data in rejection messages to a market actor that has submitted a message with invalid data.

Planned for upcoming PRs:
- [ ] Fix remaining occurrences of "TODO" in modified code
- [ ] Verify manually that rejections peeked from MessageHub contains correct error code and text
- [ ] Verify manually that all error texts and codes are correct

## References

* #922

### CODE REVIEW

**Review by HenrikSommer — APPROVED** [review]

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Infrastructure.Core/Cim/ErrorTextAttribute.cs [hunk 1]
```diff
@@ -12,23 +12,18 @@
...
 // limitations under the License.
 
-namespace GreenEnergyHub.Charges.Domain.Dtos.ChargeCommands.Validation
+using System;
+
+namespace GreenEnergyHub.Charges.Infrastructure.Core.Cim
 {
-    public class ValidationErrorMessageParameter
+    [AttributeUsage(AttributeTargets.Field)]
+    public sealed class ErrorTextAttribute : Attribute
```
Comment (BjarkeMeier) [inline source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Infrastructure.Core/Cim/ErrorTextAttribute.cs:20]: I think this should be removed.

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Infrastructure.Core/Cim/ReasonCode.cs [hunk 2]
```diff
@@ -0,0 +1,202 @@
...
+{
+    /// <summary>
+    /// All reason (error) codes and error texts origins from an Energinet internal document.
+    /// This is according to story https://github.com/Energinet-DataHub/geh-charges/issues/922.
+    ///
+    /// IMPORTANT: The name of the enum field names are used as error code in CIM documents.
+    /// </summary>
+    public enum ReasonCode
+    {
+        [ErrorText("The document is approved")]
```
Comment (BjarkeMeier) [inline source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Infrastructure.Core/Cim/ReasonCode.cs:25]: All the error text attributes should probably be removed.

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Infrastructure.Core/Cim/ReasonCodeExtensions.cs [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+using System;
+
+namespace GreenEnergyHub.Charges.Infrastructure.Core.Cim
+{
+    public static class ReasonCodeExtensions
```
Comment (BjarkeMeier) [inline source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Infrastructure.Core/Cim/ReasonCodeExtensions.cs:19]: Should probably be removed.

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Infrastructure.Core/Cim/ValidationErrors/CimValidationErrorTextTemplateMessages.cs [hunk 4]
```diff
@@ -0,0 +1,227 @@
...
+using GreenEnergyHub.Charges.Domain.Dtos.ChargeCommands.Validation;
+
+namespace GreenEnergyHub.Charges.Infrastructure.Core.Cim.ValidationErrors
+{
+    public static class CimValidationErrorTextTemplateMessages
+    {
+        /// <summary>
+        /// Errormessage for <see cref="ValidationRuleIdentifier.StartDateValidation"/>
+        /// using <see cref="CimValidationErrorTextToken.ChargeStartDateTime"/>
+        /// </summary>
```
Comment (BjarkeMeier) [inline source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Infrastructure.Core/Cim/ValidationErrors/CimValidationErrorTextTemplateMessages.cs:24]: Just a thought: How about making these relationships stronger by e.g. applying them as attributes here and then e.g. finding the template text by attribute instead of maintaining a huge switch somewhere?

Perhaps something like:
```suggestion
        [ErrormessageFor(ValidationRuleIdentifier.StartDateValidation)]
        [Tokens(CimValidationErrorTextToken.ChargeStartDateTime)]
```

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Infrastructure.Core/Cim/ValidationErrors/CimValidationErrorTextToken.cs [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+namespace GreenEnergyHub.Charges.Infrastructure.Core.Cim.ValidationErrors
+{
+    public enum CimValidationErrorTextToken
+    {
+        ChargeDescription = 1,
```
Comment (BjarkeMeier) [inline source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Infrastructure.Core/Cim/ValidationErrors/CimValidationErrorTextToken.cs:19]: Do you need the numbers? (It's easier to maintain if we can leave them out)

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Infrastructure/Contracts/Internal/ChargeCommandRejected/ChargeCommandRejected.proto [hunk 6]
```diff
@@ -67,25 +67,58 @@ message ChargeOperationContract {
...
 {
-    int32 position = 1; // The position of the price in the price list it was delivered.
-    double price = 2;
-    google.protobuf.Timestamp time = 3; // A point in time, where the price applies.
+  int32 position = 1; // The position of the price in the price list it was delivered.
+  double price = 2;
+  google.protobuf.Timestamp time = 3; // A point in time, where the price applies.
+}
+
+enum ValidationRuleIdentifierContract
```
Comment (BjarkeMeier) [inline source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Infrastructure/Contracts/Internal/ChargeCommandRejected/ChargeCommandRejected.proto:75]: Just a thought: Seems like something we could do simpler and with less effort. As the domain is a single deployment unit and this is an internal event then the enum values will be identical for sender and receiver. So how about just transferring an int and casting on both sides? We could do that for all enums in internal contracts saving some repetitive work and reducing maintenance costs.

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.MessageHub/Infrastructure/Cim/Bundles/ChargeReceipt/CimChargeReceiptConstants.cs [hunk 7]
```diff
@@ -48,5 +48,7 @@ internal static class CimChargeReceiptConstants
         internal const string ReasonCode = "code"; // Note: There are two reason codes in the CIM document
 
         internal const string ReasonText = "text";
+
+        internal const string ReasonDescription = "description";
```
Comment (BjarkeMeier) [inline source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.MessageHub/Infrastructure/Cim/Bundles/ChargeReceipt/CimChargeReceiptConstants.cs:52]: Remove?

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Tests/MessageHub/Application/Charges/AvailableChargeRejectionDataFactoryTests.cs [hunk 8]
```diff
@@ -1,71 +0,0 @@
...
-using GreenEnergyHub.Charges.TestCore.Attributes;
-using Moq;
-using NodaTime;
-using Xunit;
-using Xunit.Categories;
-
-namespace GreenEnergyHub.Charges.Tests.MessageHub.Application.Charges
-{
-    [UnitTest]
-    public class AvailableChargeRejectionDataFactoryTests
```
Comment (BjarkeMeier) [inline source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Tests/MessageHub/Application/Charges/AvailableChargeRejectionDataFactoryTests.cs:33]: Why is this deleted?

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Tests/MessageHub/Models/AvailableChargeReceiptData/CimValidationErrorDescriptionTests.cs [hunk 9]
```diff
@@ -0,0 +1,71 @@
...
+using GreenEnergyHub.Charges.Infrastructure.Core.Cim.ValidationErrors;
+using GreenEnergyHub.Charges.MessageHub.Models.AvailableChargeReceiptData;
+using GreenEnergyHub.Charges.TestCore.Attributes;
+using Xunit;
+using Xunit.Categories;
+
+namespace GreenEnergyHub.Charges.Tests.MessageHub.Models.AvailableChargeReceiptData
+{
+    [UnitTest]
+    public class CimValidationErrorTextFactoryTests
```
Comment (BjarkeMeier) [inline source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Tests/MessageHub/Models/AvailableChargeReceiptData/CimValidationErrorDescriptionTests.cs:28]: File name should match class name.

**Review by BjarkeMeier — COMMENTED** [review]
LGTM. Left some minor comments.

**Comment by BjarkeMeier** [comment]
@HenrikSommer, perhaps you should update the PR description to reflect the current state and planned work (upcoming PR).

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.MessageHub/Models/AvailableChargeReceiptData/CimValidationErrorTextFactory.cs [hunk 10]
```diff
@@ -0,0 +1,111 @@
...
+    public class CimValidationErrorTextFactory : ICimValidationErrorTextFactory
+    {
+        private readonly ICimValidationErrorTextProvider _cimValidationErrorTextProvider;
+
+        public CimValidationErrorTextFactory(ICimValidationErrorTextProvider cimValidationErrorTextProvider)
+        {
+            _cimValidationErrorTextProvider = cimValidationErrorTextProvider;
+        }
+
+        public string Create(ValidationRuleIdentifier validationRuleIdentifier, ChargeCommand chargeCommand)
```
[the remark this answers is not in the document]
Comment (HenrikSommer) [reply source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.MessageHub/Models/AvailableChargeReceiptData/CimValidationErrorTextFactory.cs:34]: Yes, done by `Create_WhenThreeMergeFields_ReturnsExpectedDescription` and `Create_MergesAllMergeFields` in `CimValidationErrorTextFactoryTests`.

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Infrastructure.Core/Cim/ErrorTextAttribute.cs (hunk 1 above)
Comment (HenrikSommer) [reply source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Infrastructure.Core/Cim/ErrorTextAttribute.cs:20]: I will reuse it for text templates as per your comment to `CimValidationErrorTextTemplateMessages`

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Tests/MessageHub/Application/Charges/AvailableChargeRejectionDataFactoryTests.cs (hunk 8 above)
Comment (HenrikSommer) [reply source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Tests/MessageHub/Application/Charges/AvailableChargeRejectionDataFactoryTests.cs:33]: I am really puzzled by this! - I have only modified it to work with our new codes and texts.

Comment (HenrikSommer) [reply source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Tests/MessageHub/Application/Charges/AvailableChargeRejectionDataFactoryTests.cs:33]: It has been moved to [source/GreenEnergyHub.Charges.Tests/MessageHub/Models/AvailableChargeReceiptData/AvailableChargeRejectionDataFactoryTests.cs](
https://github.com/Energinet-DataHub/geh-charges/blob/922_Add_parametrized_error_messages_in_rejection/source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Tests/MessageHub/Models/AvailableChargeReceiptData/AvailableChargeRejectionDataFactoryTests.cs)
